Added system_metric_collector tool

ahcorde opened the pull request:
This tool allows to create statistics based on the name of the process and arguments:

 - CPU usage (%)
 - CPU memory virtual, ResidentAnonymousMemory and physical
 - Process usage (%)
 - Process memory: a) virtual, b) resident anonymous memory and c) physical

These are the argument to launch the tool:

```
ros2 run buildfarm_perf_tests system_metric_collector -h
Options:
  -h [ --help ]           Help screen
  --timeout arg (=30)     Test duration
  --log arg (=out.csv)    Log filename
  --process_name arg      process_name
  --process_arguments arg process_arguments
```

At some point potencially this tool shoul be replace by https://github.com/ros-tooling/system_metrics_collector/

cottsay commented:
I think there was a concious decision to avoid Boost in ROS 2 entirely. No packages in `ros2.repos` require Boost at all. @tfoote - can you confirm?

If we should be avoiding Boost, is there an alternative library preferred for CLA parsing?

cottsay commented on CMakeLists.txt:
```diff
@@ -1,8 +1,42 @@
+
+include_directories(include)
+
+add_executable(system_metric_collector
```
Does this executable work on macOS?

Whatever the case, we should add some logic here so that we don't try to build it on platforms for which it isn't supported.

ahcorde replied:
`perf_test` is not compiling in MacOS, should we put some effort in to make this compatible with MacOS?

cottsay replied:
I don't think we need to make it compatible, but we should make an effort to respond appropriately if an attempt is made to compile on macOS or Windows.

There are a lot of examples where we just "skip" the parts of a package that are unsupported on a platform. It would be great to see either that or a CMake error message stating that it isn't supported.

ahcorde commented:
added "skip" for unsupported platforms

ahcorde commented:
Added details about `system_metric_collector` https://github.com/ros2/buildfarm_perf_tests/pull/12/commits/4af224555d30f6f37a115bdc7326b3365bbcb003

cottsay commented:
Just so I understand - the `system_metrics_collector` is launch independently from the process under test? And `system_metrics_collector` locates the process under test by looking for processes with a matching name and arguments?

Why don't we use the process ID instead? Couldn't we create a launch substitution that passes the PID of the process under test as a command line parameter to the `system_metrics_collector`? That would eliminate the possibility of a process name/argument collision.

Also, if we're using the `ROS2` COMM type, we're using the `RMW_IMPLEMENTATION` variable to select the RMW, so the process name and command line arguments are the same.

Alternatively, couldn't we just make the `system_metrics_collector` invoke the process under test as a subprocess?

ahcorde commented:
> Just so I understand - the system_metrics_collector is launch independently from the process under test? And system_metrics_collector locates the process under test by looking for processes with a matching name and arguments?

Exactly

> Why don't we use the process ID instead? Couldn't we create a launch substitution that passes the PID of the process under test as a command line parameter to the system_metrics_collector? That would eliminate the possibility of a process name/argument collision.

Totally agree, but I don't know how to get the PID of the process within the launch file

> Also, if we're using the ROS2 COMM type, we're using the RMW_IMPLEMENTATION variable to select the RMW, so the process name and command line arguments are the same.

if performance test solve the issue with the skiped tests we should use the COMM type

> Alternatively, couldn't we just make the system_metrics_collector invoke the process under test as a subprocess?

I was trying to avoid as much as possible the use of external code inside the ros code to avoid issues with possible memory leaks in this tools

@cottsay  does it make sense?

cottsay commented:
> Totally agree, but I don't know how to get the PID of the process within the launch file

Okay, here is a demo I wrote you: https://gist.github.com/cottsay/afb34d329e93f8f18e645b46634614cc

ahcorde commented:
Thank for the gist @cottsay :wink: 

Updated:

```
ros2 run buildfarm_perf_tests system_metric_collector -h
Options:
  -h [ --help ]           Help screen
  --timeout arg (=30)     Test duration
  --log arg (=out.csv)    Log filename
  --process_pid arg      process pid
```
Updated the others PRs with the new argument `--process_pid`:

 - https://github.com/ros2/buildfarm_perf_tests/pull/15
 - https://github.com/ros2/buildfarm_perf_tests/pull/14
 - https://github.com/ros2/buildfarm_perf_tests/pull/13

cottsay commented on src/linux_cpu_system_measurement.cpp:
```diff
@@ -0,0 +1,122 @@
+{
+  initCPUProcess();
+  makeReading();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Sleeping in a constructor seems like a really bad idea.

If you move the sleep in `main()` to be the first thing in your `while(true)` and eliminate the sleep here, it should have the same behavior, right?

ahcorde replied:
this sleep is to have a first good cpu system reading

ahcorde commented:
I have tried with /proc/<pid>/stat and it's not working. I need to reopen the file each time
